File: qpid/cpp/test/client/client_test.cpp
/*
 *
 * Copyright (c) 2006 The Apache Software Foundation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
#include <iostream>

#include <qpid/QpidError.h>
#include <qpid/client/Channel.h>
#include <qpid/client/Connection.h>
#include <qpid/client/Message.h>
#include <qpid/client/MessageListener.h>
#include <qpid/sys/Monitor.h>
#include <qpid/framing/FieldTable.h>

using namespace qpid::client;
using namespace qpid::sys;

class SimpleListener : public virtual MessageListener{
    Monitor* monitor;

public:
    inline SimpleListener(Monitor* _monitor) : monitor(_monitor){}

    inline virtual void received(Message& /*msg*/){
	std::cout << "Received message " /**<< msg **/<< std::endl;
	monitor->notify();
    }
};

int main(int argc, char**)
{
    try{               
	Connection con(argc > 1);
	Channel channel;
	Exchange exchange("MyExchange", Exchange::TOPIC_EXCHANGE);
	Queue queue("MyQueue", true);
	
	string host("localhost");
	
	con.open(host);
	std::cout << "Opened connection." << std::endl;
	con.openChannel(&channel);
	std::cout << "Opened channel." << std::endl;	
	channel.declareExchange(exchange);
	std::cout << "Declared exchange." << std::endl;
	channel.declareQueue(queue);
	std::cout << "Declared queue." << std::endl;
	qpid::framing::FieldTable args;
	channel.bind(exchange, queue, "MyTopic", args);
	std::cout << "Bound queue to exchange." << std::endl;

	//set up a message listener
	Monitor monitor;
	SimpleListener listener(&monitor);
	string tag("MyTag");
	channel.consume(queue, tag, &listener);
	channel.start();
	std::cout << "Registered consumer." << std::endl;

	Message msg;
	string data("MyMessage");
	msg.setData(data);
	channel.publish(msg, exchange, "MyTopic");
	std::cout << "Published message." << std::endl;

	{
            Monitor::ScopedLock l(monitor);
            monitor.wait();
        }
        
	con.closeChannel(&channel);
	std::cout << "Closed channel." << std::endl;
	con.close();	
	std::cout << "Closed connection." << std::endl;
    }catch(qpid::QpidError error){
	std::cout << "Error [" << error.code << "] " << error.msg << " (" << error.file << ":" << error.line << ")" << std::endl;
	return 1;
    }
    return 0;
}

File: qpid/java/perftests/src/test/java/org/apache/qpid/disttest/client/ClientTest.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 *
 */
package org.apache.qpid.disttest.client;

import static org.mockito.Matchers.isA;
import static org.mockito.Mockito.inOrder;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import java.util.Collections;
import java.util.Timer;
import java.util.TimerTask;

import org.apache.qpid.disttest.jms.ClientJmsDelegate;
import org.apache.qpid.disttest.message.Command;
import org.apache.qpid.disttest.message.ParticipantResult;
import org.apache.qpid.disttest.message.Response;
import org.apache.qpid.disttest.message.StopClientCommand;
import org.apache.qpid.test.utils.QpidTestCase;
import org.mockito.InOrder;
import org.mockito.Mockito;

public class ClientTest extends QpidTestCase
{
    private Client _client;
    private ClientJmsDelegate _delegate;
    private ClientCommandVisitor _visitor;
    private ParticipantExecutor _participant;
    private ParticipantExecutorRegistry _participantRegistry;

    @Override
    protected void setUp() throws Exception
    {
        super.setUp();
        _delegate = mock(ClientJmsDelegate.class);
        _visitor = mock(ClientCommandVisitor.class);
        _client = new Client(_delegate);
        _client.setClientCommandVisitor(_visitor);
        _participant = mock(ParticipantExecutor.class);
        when(_participant.getParticipantName()).thenReturn("testParticipantMock");

        _participantRegistry = mock(ParticipantExecutorRegistry.class);
        when(_participantRegistry.executors()).thenReturn(Collections.singletonList(_participant));
        _client.setParticipantRegistry(_participantRegistry);
    }

    public void testInitialState() throws Exception
    {
        assertEquals("Expected client to be in CREATED state", ClientState.CREATED, _client.getState());
    }

    public void testStart() throws Exception
    {
        _client.start();
        final InOrder inOrder = inOrder(_delegate);
        inOrder.verify(_delegate).setInstructionListener(_client);
        inOrder.verify(_delegate).sendRegistrationMessage();
        assertEquals("Expected client to be in STARTED state", ClientState.READY, _client.getState());
    }

    public void testStopClient() throws Exception
    {
        _client.stop();

        assertEquals("Expected client to be in STOPPED state", ClientState.STOPPED, _client.getState());
    }

    public void testProcessInstructionVisitsCommandAndResponds() throws Exception
    {
        // has to be declared to be of supertype Command otherwise Mockito verify()
        // refers to wrong method
        final Command command = new StopClientCommand();
        _client.processInstruction(command);

        verify(_visitor).visit(command);
        verify(_delegate).sendResponseMessage(isA(Response.class));
    }

    public void testWaitUntilStopped() throws Exception
    {
        stopClientLater(500);
        _client.waitUntilStopped(1000);
        verify(_delegate).destroy();
    }

    public void testStartTest() throws Exception
    {
        _client.start();
        _client.addParticipantExecutor(_participant);

        verify(_participantRegistry).add(_participant);

        _client.startTest();

        InOrder inOrder = Mockito.inOrder(_delegate, _participant);
        inOrder.verify(_delegate).startConnections();
        inOrder.verify(_participant).start(_client);
    }

    public void testTearDownTest() throws Exception
    {
        // before we can tear down the test the client needs to be in the "running test" state, which requires a participant
        _client.start();
        _client.addParticipantExecutor(_participant);
        _client.startTest();

        _client.tearDownTest();

        verify(_delegate).tearDownTest();

        verify(_participantRegistry).clear();
    }

    public void testResults() throws Exception
    {
        ParticipantResult testResult = mock(ParticipantResult.class);
        _client.sendResults(testResult);
        verify(_delegate).sendResponseMessage(testResult);
    }

    private void stopClientLater(long delay)
    {
        doLater(new TimerTask()
        {
            @Override
            public void run()
            {
                _client.stop();
            }

        }, delay);
    }

    private void doLater(TimerTask task, long delayInMillis)
    {
        Timer timer = new Timer();
        timer.schedule(task, delayInMillis);
    }

}
